// @ionic/cli/lib/project/vue/build.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chalk = require("chalk");
const build_1 = require("../../build");
const errors_1 = require("../../errors");
class VueBuildRunner extends build_1.BuildRunner {
    constructor(e) {
        super();
        this.e = e;
    }
    async getCommandMetadata() {
        return {};
    }
    createOptionsFromCommandLine(inputs, options) {
        const baseOptions = super.createBaseOptionsFromCommandLine(inputs, options);
        return {
            ...baseOptions,
            type: 'vue',
        };
    }
    async buildProject(options) {
        const cli = this.getPkgManagerBuildCLI();
        if (!await cli.resolveScript()) {
            throw new errors_1.RunnerException(`Cannot perform build.\n` +
                `Since you're using the ${chalk.bold('Vue')} project type, you must provide the ${chalk.green(cli.script)} npm script so the Ionic CLI can build your project.`);
        }
        await cli.build(options);
    }
}
exports.VueBuildRunner = VueBuildRunner;
